compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/codegen/JvmObservabilityGen.java

```diff
@@ -170,6 +193,65 @@ void instrumentPackage(BIRPackage pkg) {
         }
     }
 
+    /**
+     * Adding Java Interop calls to basic blocks.
+     * Here the JI calls are added for all kinds of terminators.
+     *
+     * First we check if there are position details for instructions, if present we add the JI calls with those
+     * positions else, we consider the terminator position to create the JI call.
+     *
+     * @param func The function of which the instructions should be rewritten
+     * @param pkg The package containing the function
+     */
+    private void rewriteControlFlowInvocation(BIRFunction func, BIRPackage pkg) {
+        int i = 0;
+        while (i < func.basicBlocks.size()) {
+            // Basic blocks with JI method calls are added for all kinda of Terminators
+            BIRBasicBlock startBB = func.basicBlocks.get(i);
+            Location desugaredPos;
+            // First we give the priority to Instructions,
+            // If no instructions are found, then we get the Terminator position
+            if (startBB.instructions.size() != 0) {
+                desugaredPos = startBB.instructions.get(0).pos;
+            } else {
+                desugaredPos = startBB.terminator.pos;
+            }
+            if (desugaredPos != null) {
+                BIRBasicBlock newBB = insertBasicBlock(func, i + 1);
+                swapBasicBlockContent(startBB, newBB);
+                injectCheckpointCall(startBB, pkg, desugaredPos);
+                startBB.terminator.thenBB = newBB;
+                //Fix error entries in the error entry table
+                fixErrorTable(func, startBB, newBB);
+                i += 1; // Number of inserted BBs
+            }
+            i += 1;
+        }
+    }
+
+    /**
+     * Inject checkpoint JI method call to a basic block.
+     * @param startBB The basic block to which the checkpoint call should be injected
+     * @param pkg The package the invocation belongs to
+     * @param desugaredInsPosition The source code position of the invocation
+     */
+    private void injectCheckpointCall(BIRBasicBlock startBB, BIRPackage pkg, Location desugaredInsPosition) {
```

Comment:
Shall we rename `desugaredInsPosition` to `originalInsPosition` ? Both are the same here, but we switch names halfway in this function method body.